File: takengine/takengine/src/main/java/com/atakmap/map/layer/raster/tilereader/opengl/NodeContextResources.java
package com.atakmap.map.layer.raster.tilereader.opengl;

import android.graphics.PointF;
import android.opengl.GLES30;

import com.atakmap.lang.Unsafe;
import com.atakmap.opengl.GLTexture;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.FloatBuffer;
import java.nio.ShortBuffer;

final class NodeContextResources {
    final int[] uniformGriddedTexCoords = new int[GLQuadTileNode3.MAX_GRID_SIZE];
    final int[] uniformGriddedIndices = new int[GLQuadTileNode3.MAX_GRID_SIZE];
    final ByteBuffer coordStreamBuffer;
    final FloatBuffer coordStreamBufferF;
    final ShortBuffer coordStreamBufferS;
    final static int[] discardBuffers = new int[16];
    static int numNumDiscardBuffers = 0;

    NodeContextResources() {
        // large enough to hold all mesh indices or vertex data or interleaved quad for texture-copy (position.xy+texcoord.uv)
        coordStreamBuffer = Unsafe.allocateDirect(
                Math.max(
                        Math.max(
                                // max indices
                                GLTexture.getNumQuadMeshIndices(GLQuadTileNode3.MAX_GRID_SIZE, GLQuadTileNode3.MAX_GRID_SIZE) * 2,
                                // max pos/texcoord
                                GLTexture.getNumQuadMeshVertices(GLQuadTileNode3.MAX_GRID_SIZE, GLQuadTileNode3.MAX_GRID_SIZE) * 12),
                        // max interleaved pox.xy,texcoord.uv for texture copy
                        4 * (8 + 8)));
        coordStreamBuffer.order(ByteOrder.nativeOrder());
        coordStreamBufferF = coordStreamBuffer.asFloatBuffer();
        coordStreamBufferS = coordStreamBuffer.asShortBuffer();
    }

    final boolean isUniformGrid(int gridWidth, int gridHeight) {
        return (gridWidth == gridHeight) &&
               (gridWidth > 0) &&
               ((gridWidth-1) < uniformGriddedTexCoords.length);
    }
    int getUniformGriddedTexCoords(int gridSize) {
        int uniformTexCoords = uniformGriddedTexCoords[gridSize-1];
        if(uniformTexCoords == GLES30.GL_NONE) {
            final int numVerts = GLTexture.getNumQuadMeshVertices(gridSize, gridSize);
            coordStreamBuffer.clear();
            coordStreamBufferF.clear();
            GLTexture.createQuadMeshTexCoords(new PointF(0f, 0f),
                                              new PointF(1f, 0f),
                                              new PointF(1f, 1f),
                                              new PointF(0f, 1f),
                                              gridSize,
                                              gridSize,
                                              coordStreamBufferF);

            uniformTexCoords = genBuffer();
            GLES30.glBindBuffer(GLES30.GL_ARRAY_BUFFER, uniformTexCoords);
            GLES30.glBufferData(GLES30.GL_ARRAY_BUFFER, numVerts*2*4, coordStreamBuffer, GLES30.GL_STATIC_DRAW);
            GLES30.glBindBuffer(GLES30.GL_ARRAY_BUFFER, GLES30.GL_NONE);
            uniformGriddedTexCoords[gridSize-1] = uniformTexCoords;
        }
        return uniformTexCoords;
    }
    int getUniformGriddedIndices(int gridSize) {
        int uniformIndices = uniformGriddedIndices[gridSize-1];
        if(uniformIndices == GLES30.GL_NONE) {
            final int numIndices = GLTexture.getNumQuadMeshIndices(gridSize, gridSize);
            uniformIndices = genBuffer();
            coordStreamBuffer.clear();
            coordStreamBufferS.clear();
            GLTexture.createQuadMeshIndexBuffer(gridSize, gridSize, coordStreamBufferS);
            GLES30.glBindBuffer(GLES30.GL_ELEMENT_ARRAY_BUFFER, uniformIndices);
            GLES30.glBufferData(GLES30.GL_ELEMENT_ARRAY_BUFFER, numIndices * 2, coordStreamBuffer, GLES30.GL_STATIC_DRAW);
            GLES30.glBindBuffer(GLES30.GL_ELEMENT_ARRAY_BUFFER, GLES30.GL_NONE);
            uniformGriddedIndices[gridSize-1] = uniformIndices;
        }
        return uniformIndices;
    }
    int discardBuffer(int id) {
        if(id == GLES30.GL_NONE)
            return GLES30.GL_NONE;
        if(numNumDiscardBuffers == discardBuffers.length) {
            GLES30.glDeleteBuffers(numNumDiscardBuffers, discardBuffers, 0);
            numNumDiscardBuffers = 0;
        }
        discardBuffers[numNumDiscardBuffers++] = id;
        return GLES30.GL_NONE;

    }
    int genBuffer() {
        if(numNumDiscardBuffers == 0) {
            GLES30.glGenBuffers(1, discardBuffers, 0);
            numNumDiscardBuffers++;
        }
        return discardBuffers[--numNumDiscardBuffers];
    }
    void deleteBuffers() {
        if(numNumDiscardBuffers > 0)
            GLES30.glDeleteBuffers(numNumDiscardBuffers, discardBuffers, 0);
        numNumDiscardBuffers = 0;
    }
}

File: takkernel/engine/src/main/cpp/renderer/raster/tilereader/NodeContextResources.cpp
#ifdef MSVC
#include "renderer/raster/tilereader/NodeContextResources.h"

#include <algorithm>

#include "renderer/GL.h"
#include "renderer/GLTexture2.h"
#include "thread/Lock.h"
#include "thread/Mutex.h"

using namespace TAK::Engine::Renderer::Raster::TileReader;

using namespace TAK::Engine::Core;
using namespace TAK::Engine::Math;
using namespace TAK::Engine::Renderer;
using namespace TAK::Engine::Thread;
using namespace TAK::Engine::Util;

#define GLQuadTileNode3_MAX_GRID_SIZE 32

namespace
{
    struct {
        Mutex mutex;
        std::map<const RenderContext*, std::shared_ptr<NodeContextResources>> value;
    } cache;
}

NodeContextResources::NodeContextResources() NOTHROWS :
    coordStreamBuffer(std::max(
        std::max(
            // max indices
            GLTexture2_getNumQuadMeshIndices(GLQuadTileNode3_MAX_GRID_SIZE, GLQuadTileNode3_MAX_GRID_SIZE) * 2u,
            // max pos/texcoord
            GLTexture2_getNumQuadMeshVertices(GLQuadTileNode3_MAX_GRID_SIZE, GLQuadTileNode3_MAX_GRID_SIZE) * 12u),
        // max interleaved pox.xy,texcoord.uv for texture copy
        (std::size_t)(4u * (8u + 8u)))
    ),
    coordStreamBufferF(reinterpret_cast<float *>(const_cast<uint8_t *>(coordStreamBuffer.get())), coordStreamBuffer.size()/4u),
    coordStreamBufferS(reinterpret_cast<uint16_t *>(const_cast<uint8_t *>(coordStreamBuffer.get())), coordStreamBuffer.size()/2u)
{
    for (std::size_t i = 0u; i < GLQuadTileNode3_MAX_GRID_SIZE; i++) {
        uniformGriddedIndices[i] = GL_NONE;
        uniformGriddedTexCoords[i] = GL_NONE;
    }
}

bool NodeContextResources::isUniformGrid(const std::size_t gridWidth, const std::size_t gridHeight) const NOTHROWS
{
    return (gridWidth == gridHeight) &&
            (gridWidth > 0) &&
            ((gridWidth-1) < GLQuadTileNode3_MAX_GRID_SIZE);
}
GLuint NodeContextResources::getUniformGriddedTexCoords(const std::size_t gridSize) NOTHROWS
{
    if (gridSize >= GLQuadTileNode3_MAX_GRID_SIZE)
        return GL_NONE;
    GLuint uniformTexCoords = uniformGriddedTexCoords[gridSize-1];
    if(uniformTexCoords == GL_NONE) {
        const std::size_t numVerts = GLTexture2_getNumQuadMeshVertices(gridSize, gridSize);
        coordStreamBuffer.reset();
        coordStreamBufferF.reset();
        GLTexture2_createQuadMeshTexCoords(coordStreamBufferF,
                                            Point2<float>(0.f, 0.f),
                                            Point2<float>(1.f, 0.f),
                                            Point2<float>(1.f, 1.f),
                                            Point2<float>(0.f, 1.f),
                                            gridSize,
                                            gridSize);

        uniformTexCoords = genBuffer();
        glBindBuffer(GL_ARRAY_BUFFER, uniformTexCoords);
        glBufferData(GL_ARRAY_BUFFER, numVerts*2*4, coordStreamBuffer.get(), GL_STATIC_DRAW);
        glBindBuffer(GL_ARRAY_BUFFER, GL_NONE);
        uniformGriddedTexCoords[gridSize-1] = uniformTexCoords;
    }
    return uniformTexCoords;
}
GLuint NodeContextResources::getUniformGriddedIndices(const std::size_t gridSize) NOTHROWS
{
    int uniformIndices = uniformGriddedIndices[gridSize-1];
    if(uniformIndices == GL_NONE) {
        const std::size_t numIndices = GLTexture2_getNumQuadMeshIndices(gridSize, gridSize);
        uniformIndices = genBuffer();
        coordStreamBuffer.reset();
        coordStreamBufferS.reset();
        GLTexture2_createQuadMeshIndexBuffer(coordStreamBufferS, GL_UNSIGNED_SHORT, gridSize, gridSize);
        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, uniformIndices);
        glBufferData(GL_ELEMENT_ARRAY_BUFFER, numIndices * 2, coordStreamBuffer.get(), GL_STATIC_DRAW);
        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, GL_NONE);
        uniformGriddedIndices[gridSize-1] = uniformIndices;
    }
    return uniformIndices;
}
GLuint NodeContextResources::discardBuffer(const GLuint id) NOTHROWS
{
    if(id == GL_NONE)
        return GL_NONE;
    if(numNumDiscardBuffers == 16u) {
        glDeleteBuffers(numNumDiscardBuffers, discardBuffers);
        numNumDiscardBuffers = 0;
    }
    discardBuffers[numNumDiscardBuffers++] = id;
    return GL_NONE;

}
GLuint NodeContextResources::genBuffer() NOTHROWS
{
    // if no buffers are available, generate one
    if(numNumDiscardBuffers == 0) {
        glGenBuffers(1, discardBuffers);
        numNumDiscardBuffers++;
    }
    // pull buffer off of end of discard list
    return discardBuffers[--numNumDiscardBuffers];
}
void NodeContextResources::deleteBuffers() NOTHROWS
{
    // delete all discard buffers
    if(numNumDiscardBuffers > 0)
        glDeleteBuffers(numNumDiscardBuffers, discardBuffers);
    numNumDiscardBuffers = 0;
}

TAKErr TAK::Engine::Renderer::Raster::TileReader::NodeContextResources_get(std::shared_ptr<NodeContextResources>& value, const RenderContext &ctx) NOTHROWS
{
    TAKErr code(TE_Ok);
    Lock lock(cache.mutex);
    TE_CHECKRETURN_CODE(lock.status);
    auto entry = cache.value.find(&ctx);
    // check cache
    if (entry != cache.value.end()) {
        value = entry->second;
        return code;
    }

    // create a new `NodeContextResources` and add to cache
    value = std::shared_ptr<NodeContextResources>(new NodeContextResources());
    cache.value[&ctx] = value;
    return code;
}
#endif